src/typechecker.rs

```diff
@@ -626,10 +626,46 @@ impl<'a> Typechecker<'a> {
         &mut self,
         name: NodeId,
         params: NodeId,
-        _return_ty: Option<NodeId>,
+        return_ty: Option<NodeId>,
         block: NodeId,
         node_id: NodeId,
     ) {
+        let return_ty = return_ty
+            .map(|ty| {
+                let AstNode::InOutTypes(types) = self.compiler.get_node(ty) else {
+                    panic!("internal error: return type is not a return type");
```

Comment:
This pattern is repeated a lot and is kind of unwieldy. It could use some helper or API refactor. Just a note for the future.

Reply:
Yeah. I was thinking about a macro

Reply:
I was thinking along the lines that if we made each AST node its own type, we could use generics, similar to [this solution](https://stackoverflow.com/a/34954128). That would also allow us to deduplicate all the `.is_<token>()` and `.<token>()` methods of the Parser.

I'm not a fan of macros since they have different syntax from the rest of Rust, and they can make code hard to read, but sometimes they're helpful.